text/0011-gem-checksum-verification.md

```diff
@@ -0,0 +1,184 @@
+
+For users that are not specifically looking for the feature, common commands like `bundle install`, `bundle update`, `bundle lock`, `bundle add`, should automatically make use of checksums when available. When there are missing checksums that must be fetched with a specific interaction, the UI should say what that action is.
+
+Example:
+
+```
+$ bundle install
+Bundle complete! 88 Gemfile dependencies, 256 gems now installed.
+Use `bundle info [gemname]` to see where a bundled gem is installed.
+256 gems are missing checksums. Source integrity could not be verified.
+Use `bundle lock` to add checksums to Gemfile.lock.
+```
+
+When a gem version is updated in the lockfile using a source that has checksums (like rubygems.org), the checksum should be recorded. Running `bundle update` or `bundle add` should record the checksum from the source into the Gemfile.lock. If a checksum is not available from the source because the source does not provide such info, then a checksum should be created during install using the .gem file or a warning should be printed.
+
+If a user wishes to ensure that the environment they are using contains only gems installed from verified sources that match the checksums in the Gemfile.lock, the command `bundle pristine` should be run. The `bundle pristine` command installs every gem fresh from the .gem source, but does not fetch gems from the server or read the source index. Therefore, if a users wants all gems to be verified upon install, `bundle pristine` should be run after checksums are written to the Gemfile.lock.
+
+Example:
+
+```
+$ bundle pristine
+Installing rake 13.0.6
+…
+Installing rspec 3.12.0
+42 gems are missing checksums. Source integrity could not be verified.
+Use `bundle lock` to add checksums to Gemfile.lock.
+$ bundle lock --update
+Writing lockfile to path/to/Gemfile.lock
+$ bundle pristine
+Installing rake 13.0.6 (verified)
+…
+Installing rspec 3.12.0 (verified)
+$
+```
+
+If a gem that is about to be installed does not match the checksum, an error is generated.
+
+Example:
+
+```
+$ bundle install
+Installing rake 13.0.6
+Bundler cannot continue installing rake-13.0.6.
+The checksum for the downloaded `rake-13.0.6.gem` does not match the known checksum for the gem. 
+
+Expected: sha256-814828c34f1315d7e7b7e8295184577cc4e969bad6156ac069d02d63f58d82e8 from: Gemfile.lock:21:1 CHECKSUMS rake (13.0.6)
+Downloaded: sha256-69b69b69b69b69b69b69b69b69b69b69b69b69b69b69b69b69b69b69b69b69b6 from:
+path/to/rake-13.0.6.gem SHA256 hexdigest
+
+To resolve this issue:
+Delete the downloaded gem files referenced above and run `bundle install`
+
+If you are sure that the downloaded gem is correct, you can run…
+(RFC TODO: Currently undecided on what the command should be here to allow unverified gems.)
```

Comment:
Yeah, I think `gem "foo", checksum_override: "abc123"` is probably the most reasonable way to do this... if we actually want to allow it? I'm not sure if we do.

Reply:
it would have to be a hash, because each platform would have its own checksum.

Reply:
The error could output the command to run to ignore this specific gem's mismatch and then we would just record it in the gemfile.lock. I imagine just changing the relevant line in the CHECKSUMS section to `rake (1.0.0) IGNORED`. Then since it's transiently in the lockfile, the next update to a different version would either produce a new error or resolve correctly.

Reply:
Should we entirely remove the message about disabling the feature and just update the docs?

This would be in addition to adding a command for ignoring a single checksum mismatch, if we decide to do that.

Reply:
What would be the use case for needing this? Could we instead allow users to vendor the specific gem with this checksum? Having random hardcoded checksums in the Gemfile feels weird. Everybody will need to have access to the source `.gem` for it to work, right? So how come would the mismatch happen?

Maybe alternatively (if I understand correctly how this feature would play with vendoring) we could allow users to vendor specific `.gem` files? I think this would be similar to what was requested at https://github.com/rubygems/rubygems/issues/4495?